subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -115,180 +84,88 @@ public ClassSetDiff findChangesSince(ClassSetAnalysis other) {
+     * Otherwise follows the below rules for all of the given classes, as well as the classes that were marked as "always recompile" by annotation processing:
+     *
+     * Starts at this class and capture all classes that reference this class and all classes and resources that were generated from this class.
+     * Then does the same analysis for all classes that expose this class on their ABI recursively until no more new classes are discovered.
      */
-    public DependentsSet findTransitiveDependents(Collection<String> classes, Map<String, IntSet> constants) {
+    public DependentsSet findTransitiveDependents(Collection<String> classes, Map<String, IntSet> changedConstantsByClass) {
         if (classes.isEmpty()) {
             return DependentsSet.empty();
         }
-        final Set<String> accessibleResultClasses = new HashSet<>();
-        final Set<String> privateResultClasses = new HashSet<>();
-        final Set<GeneratedResource> resultResources = new HashSet<>();
-        for (String cls : classes) {
-            DependentsSet d = findTransitiveDependents(cls, constants.getOrDefault(cls, IntSets.emptySet()));
-            if (d.isDependencyToAll()) {
-                return d;
-            }
-            if (d.isEmpty()) {
-                continue;
-            }
-            Set<String> accessibleDependentClasses = d.getAccessibleDependentClasses();
-            Set<String> privateDependentClasses = d.getPrivateDependentClasses();
-            Set<GeneratedResource> dependentResources = d.getDependentResources();
-
-            accessibleResultClasses.addAll(accessibleDependentClasses);
-            privateResultClasses.addAll(privateDependentClasses);
-            resultResources.addAll(dependentResources);
-        }
-        return DependentsSet.dependents(privateResultClasses, accessibleResultClasses, resultResources);
-    }
-
-    /**
-     * Computes the transitive dependents of a changed class.
-     * If the class had any changes to inlineable constants, these need to be provided as the second parameter.
-     */
-    public DependentsSet findTransitiveDependents(String className, IntSet constants) {
         String fullRebuildCause = annotationProcessingData.getFullRebuildCause();
         if (fullRebuildCause != null) {
             return DependentsSet.dependencyToAll(fullRebuildCause);
         }
-
-        DependentsSet deps = getDependents(className);
-        if (deps.isDependencyToAll()) {
-            return deps;
-        }
-        if (!constants.isEmpty() && !compilerApiData.isSupportsConstantsMapping()) {
-            return DependentsSet.dependencyToAll("an inlineable constant in '" + className + "' has changed");
-        }
-        Set<String> classesDependingOnAllOthers = annotationProcessingData.participatesInClassGeneration(className) ? annotationProcessingData.getGeneratedTypesDependingOnAllOthers() : Collections.emptySet();
```

Comment:
There shouldn't have been any `if` here, they are called "depending on all others" for a good reason, since they can be affected by newly added files.